### Overview

**simsite** is a web-based reference vignettes for illustrating the usage of simulation methods for single-cell RNA sequencing data.

It contains three main sections:

* Programming usage. All demonstrations and programming environment are in **R**. It shows a comprehensive pipeline from installing three main R packages ([simutils](https://github.com/duohongrui/simutils), [simmethods](https://github.com/duohongrui/simmethods) and [simpipe](https://github.com/duohongrui/simpipe)) or just using [simpipeDocker](https://github.com/duohongrui/simpipe_Docker), estimating parameters from real data, simulating different kinds of new datasets and evaluating the simulated datasets.

* Online usage. It contains some methods that show the best performance and users can simulate new datasets fast and conveniently.

* References. [**simmethods**](https://github.com/duohongrui/simmethods) is the core package in our project. We collected 42+ methods and bundled them into a single package. For each method, we have already prepared a detailed vignette for users to learn to use the methods they are interested in.

#### Quick start

[Programming start](/programming/) ⬅

[Install R packages or download Docker images](/programming/1-installation) ⬅

[Online usage start](/online/) ⬅

### Features

* A comprehensive collection of simulation methods for single-cell RNA sequencing data.

* A [Docker image](https://hub.docker.com/r/duohongrui/simpipe) is provided to help users establish the simulation environment easily and users can use it in local R.

* Both local programming environment and interactive online tool can be used.

* Detailed vignettes for every simulation methods using [simmethods](https://github.com/duohongrui/simmethods) package.


### Citation
A comparison of simulation methods for single-cell RNA sequencing data.

### Help and Issues

If you need help or have any issue about our tools and vignettes. Send an email to [duohongrui](https://github.com/duohongrui) (duohongrui@cqnu.edu.cn) or raise an issue on [github](https://github.com/duohongrui).

### New Methods

We are glad to add new simulation methods (especially for simulating spatial transcriptomics data) if some methods are innovative and creative that many users commonly used. If you have the requirements, please tell Hongrui Duo by email (duohongrui@cqnu.edu.cn) or raise an [issue](https://github.com/duohongrui/simmethods/issues) for that.
