---
title: "Splat"
output:
  blogdown::html_page:
    toc: yes
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-splat" id="toc-simulating-datasets-using-splat">Simulating datasets using Splat</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
<li><a href="#simulate-two-or-more-groups" id="toc-simulate-two-or-more-groups">Simulate two or more groups</a></li>
<li><a href="#simulate-two-or-more-batches" id="toc-simulate-two-or-more-batches">Simulate two or more batches</a></li>
<li><a href="#simulate-more-groups-and-batches-simutaniously" id="toc-simulate-more-groups-and-batches-simutaniously">Simulate more groups and batches simutaniously</a></li>
<li><a href="#return-results-with-different-format" id="toc-return-results-with-different-format">Return results with different format</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/Splat.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here Splat method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data
dim(ref_data)
# [1] 4000  160</code></pre>
<p>Using <code>simmethods::Splat_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::Splat_estimation(ref_data = ref_data,
                                                verbose = T,
                                                seed = 10)
# Estimating parameters using Splat</code></pre>
</div>
<div id="simulating-datasets-using-splat" class="section level2">
<h2>Simulating datasets using Splat</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells and genes</li>
<li>Simulate two or more groups</li>
<li>Simulate two or more batches</li>
<li>Simulate more groups and batches simutaniously</li>
<li>Return results with different format</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group and one batch of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 3 columns
#         cell_name       batch       group
#       &lt;character&gt; &lt;character&gt; &lt;character&gt;
# Cell1       Cell1      Batch1      Group1
# Cell2       Cell2      Batch1      Group1
# Cell3       Cell3      Batch1      Group1
# Cell4       Cell4      Batch1      Group1
# Cell5       Cell5      Batch1      Group1
# Cell6       Cell6      Batch1      Group1</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 1 column
#         gene_name
#       &lt;character&gt;
# Gene1       Gene1
# Gene2       Gene2
# Gene3       Gene3
# Gene4       Gene4
# Gene5       Gene5
# Gene6       Gene6</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>In Splat, we can not set <code>nCells</code> directly and should set <code>batchCells</code> instead. For example, if we want to simulate 500 cells, we can type <code>other_prior = list(batchCells = 500)</code>. For genes, we can just set <code>nGenes</code>.
Here, we simulate a new dataset with 500 cells and 1000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = 500,
                     nGenes = 1000),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
</div>
<div id="simulate-two-or-more-groups" class="section level3">
<h3>Simulate two or more groups</h3>
<p>In Splat, we can not set <code>nGroups</code> directly and should set <code>prob.group</code> instead. For example, if we want to simulate 2 groups, we can type <code>other_prior = list(prob.group = c(0.5, 0.5))</code>. Note that the sum of <code>prob.group</code> numeric vector must equal to 1, so we can also set <code>prob.group = c(0.3, 0.7)</code>.</p>
<p>In addtion, if we want to simulate three or more groups, we should obey the rules:</p>
<ul>
<li>The length of <code>prob.group</code> vector must always equal to the number of groups.</li>
<li>The sum of <code>prob.group</code> numeric vector must equal to 1.</li>
</ul>
<p>For demonstration, we will simulate three groups using the learned parameters.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = 500,
                     nGenes = 1000,
                     prob.group = c(0.1, 0.3, 0.6)),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 1</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$group)
# 
# Group1 Group2 Group3 
#     46    156    298</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.1
# yes 
# 0.1</code></pre>
<p>We can see that the proportion of differential expressed genes is 0.1 (equals to the default). Next, if we want to know the fold change between two groups, we will do division with the groups that we are interested in.</p>
<pre class="r"><code>## fc between group2 and group1
fc_group1_to_group2 &lt;- gene_info$DEFacGroup2/gene_info$DEFacGroup1
## fc between group3 and group1
fc_group1_to_group3 &lt;- gene_info$DEFacGroup3/gene_info$DEFacGroup1
## fc between group3 and group2
fc_group2_to_group3 &lt;- gene_info$DEFacGroup3/gene_info$DEFacGroup2</code></pre>
</div>
<div id="simulate-two-or-more-batches" class="section level3">
<h3>Simulate two or more batches</h3>
<p>In Splat, we can not set <code>nBatches</code> directly and should set <code>batchCells</code> instead. For example, if we want to simulate 2 batches, we can type <code>other_prior = list(batchCells = c(250, 250))</code>. Note that the sum of <code>batchCells</code> numeric vector represents the total number of cells and the length of the vector equals to the number of batches.</p>
<p>In addtion, if we want to simulate three or more batches, we should obey the rules:</p>
<ul>
<li>The length of <code>batchCells</code> vector always equals to the number of batches.</li>
<li>The sum of <code>batchCells</code> numeric vector represents the total number of cells.</li>
</ul>
<p>For demonstration, we will simulate three batches using the learned parameters.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = c(200, 300),
                     nGenes = 1000),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 1
# de.prob: 0.1
# nBatches: 2</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#    200    300</code></pre>
</div>
<div id="simulate-more-groups-and-batches-simutaniously" class="section level3">
<h3>Simulate more groups and batches simutaniously</h3>
<p>As mentioned before, we can set <code>prob.group</code> and <code>batchCells</code> to determine the number of groups and batches and we can also set <code>de.prob</code> to specify the proportion of DEGs. Here, we simulate a dataset with following settings:</p>
<ul>
<li>1000 cells</li>
<li>5000 genes</li>
<li>three groups with 0.2 proportion of DEGs</li>
<li>two batches</li>
</ul>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = c(500, 500),
                     nGenes = 5000,
                     de.prob = 0.2,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 1000
# nGenes: 5000
# nGroups: 3
# de.prob: 0.2
# nBatches: 2</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 5000 1000</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#    500    500
table(cell_info$group)
# 
# Group1 Group2 Group3 
#    186    321    493</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result)
#    yes 
# 0.1932
### fc
fc_group2_to_group3 &lt;- gene_info$DEFacGroup3/gene_info$DEFacGroup2</code></pre>
</div>
<div id="return-results-with-different-format" class="section level3">
<h3>Return results with different format</h3>
<p>In simmethods package, we provide four formats of results to users without data format conversion, including <code>list</code>, <code>SingleCellExperiment</code>, <code>Seurat</code> and <code>h5ad</code>. The previous three formats are compatible with R environment and the last <code>h5ad</code> format is suitable for <strong>Python</strong> environment and can be imported by <code>scanpy.read_h5ad</code> function.</p>
<div id="list" class="section level4">
<h4>list</h4>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = c(100, 100),
                     nGenes = 1000,
                     de.prob = 0.1,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 200
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 2</code></pre>
<pre class="r"><code>str(simulate_result)
# List of 2
#  $ simulate_result   :List of 3
#   ..$ count_data: int [1:1000, 1:200] 16 18 14 40 5 23 9 694 27 21 ...
#   .. ..- attr(*, &quot;dimnames&quot;)=List of 2
#   .. .. ..$ : chr [1:1000] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; ...
#   .. .. ..$ : chr [1:200] &quot;Cell1&quot; &quot;Cell2&quot; &quot;Cell3&quot; &quot;Cell4&quot; ...
#   ..$ col_meta  :&#39;data.frame&#39;:    200 obs. of  3 variables:
#   .. ..$ cell_name: chr [1:200] &quot;Cell1&quot; &quot;Cell2&quot; &quot;Cell3&quot; &quot;Cell4&quot; ...
#   .. ..$ batch    : chr [1:200] &quot;Batch1&quot; &quot;Batch1&quot; &quot;Batch1&quot; &quot;Batch1&quot; ...
#   .. ..$ group    : Factor w/ 3 levels &quot;Group1&quot;,&quot;Group2&quot;,..: 2 2 3 2 3 3 3 2 3 3 ...
#   ..$ row_meta  :&#39;data.frame&#39;:    1000 obs. of  7 variables:
#   .. ..$ gene_name     : chr [1:1000] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; ...
#   .. ..$ de_gene       : chr [1:1000] &quot;no&quot; &quot;no&quot; &quot;no&quot; &quot;no&quot; ...
#   .. ..$ BatchFacBatch1: num [1:1000] 0.91 0.977 1.054 1.171 1.002 ...
#   .. ..$ BatchFacBatch2: num [1:1000] 1.105 0.95 0.76 0.776 0.996 ...
#   .. ..$ DEFacGroup1   : num [1:1000] 1 1 1 1 1 1 1 1 1 1 ...
#   .. ..$ DEFacGroup2   : num [1:1000] 1 1 1 1 1 1 1 1 1 1 ...
#   .. ..$ DEFacGroup3   : num [1:1000] 1 1 1 1 1 1 1 1 1 1 ...
#  $ simulate_detection:&#39;data.frame&#39;:   1 obs. of  4 variables:
#   ..$ Function_Call     : chr &quot;simulate_result&lt;-splatter::splatSimulate(parameters,method=submethod,verbose=verbose)&quot;
#   ..$ Elapsed_Time_sec  : num 0.172
#   ..$ Total_RAM_Used_MiB: num 7
#   ..$ Peak_RAM_Used_MiB : num 4781066</code></pre>
<pre class="r"><code>counts &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
head(cell_info)
#       cell_name  batch  group
# Cell1     Cell1 Batch1 Group2
# Cell2     Cell2 Batch1 Group2
# Cell3     Cell3 Batch1 Group3
# Cell4     Cell4 Batch1 Group2
# Cell5     Cell5 Batch1 Group3
# Cell6     Cell6 Batch1 Group3</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
head(gene_info)
#       gene_name de_gene BatchFacBatch1 BatchFacBatch2 DEFacGroup1 DEFacGroup2
# Gene1     Gene1      no      0.9098860      1.1054169           1           1
# Gene2     Gene2      no      0.9774161      0.9501320           1           1
# Gene3     Gene3      no      1.0541276      0.7597880           1           1
# Gene4     Gene4      no      1.1708139      0.7762219           1           1
# Gene5     Gene5      no      1.0017116      0.9963063           1           1
# Gene6     Gene6      no      0.7654326      1.1497335           1           1
#       DEFacGroup3
# Gene1           1
# Gene2           1
# Gene3           1
# Gene4           1
# Gene5           1
# Gene6           1</code></pre>
</div>
<div id="singlecellexperiment" class="section level4">
<h4>SingleCellExperiment</h4>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SingleCellExperiment&quot;,
  other_prior = list(batchCells = c(100, 100),
                     nGenes = 1000,
                     de.prob = 0.1,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 200
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 2</code></pre>
<pre class="r"><code>counts &lt;- counts(simulate_result[[&quot;simulate_result&quot;]])
## cell information
cell_info &lt;- as.data.frame(colData(simulate_result[[&quot;simulate_result&quot;]]))
head(cell_info)
#       cell_name  batch  group
# Cell1     Cell1 Batch1 Group2
# Cell2     Cell2 Batch1 Group2
# Cell3     Cell3 Batch1 Group3
# Cell4     Cell4 Batch1 Group2
# Cell5     Cell5 Batch1 Group3
# Cell6     Cell6 Batch1 Group3</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- as.data.frame(rowData(simulate_result[[&quot;simulate_result&quot;]]))
head(gene_info)
#       gene_name de_gene BatchFacBatch1 BatchFacBatch2 DEFacGroup1 DEFacGroup2
# Gene1     Gene1      no      0.9098860      1.1054169           1           1
# Gene2     Gene2      no      0.9774161      0.9501320           1           1
# Gene3     Gene3      no      1.0541276      0.7597880           1           1
# Gene4     Gene4      no      1.1708139      0.7762219           1           1
# Gene5     Gene5      no      1.0017116      0.9963063           1           1
# Gene6     Gene6      no      0.7654326      1.1497335           1           1
#       DEFacGroup3
# Gene1           1
# Gene2           1
# Gene3           1
# Gene4           1
# Gene5           1
# Gene6           1</code></pre>
</div>
<div id="seurat" class="section level4">
<h4>Seurat</h4>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;Seurat&quot;,
  other_prior = list(batchCells = c(100, 100),
                     nGenes = 1000,
                     de.prob = 0.1,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 200
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 2
seurat_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
## Overview
seurat_result
# An object of class Seurat 
# 1000 features across 200 samples within 1 assay 
# Active assay: originalexp (1000 features, 0 variable features)</code></pre>
<pre class="r"><code>## count matrix
counts &lt;- seurat_result@assays$originalexp@counts
counts[1:10, 1:10]
# 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;
#    [[ suppressing 10 column names &#39;Cell1&#39;, &#39;Cell2&#39;, &#39;Cell3&#39; ... ]]
#                                         
# Gene1   16 .   7   5  . .  .  11  220  1
# Gene2   18 .  12  10  . 1  1  32  222  2
# Gene3   14 .   4   2  1 .  .  23  101  5
# Gene4   40 1  13  18  . 1  1  27  464  5
# Gene5    5 .   2   4  . .  .   6   93  2
# Gene6   23 .   6  11  . .  .  14  334  5
# Gene7    9 .  13   2  . .  .  13  189  .
# Gene8  694 7 509 229 26 6 12 774 7932 61
# Gene9   27 .  16   6  2 1  .  12  311  .
# Gene10  21 .  13   6  . .  .  24  199  5</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- seurat_result@meta.data
head(cell_info)
#          orig.ident nCount_originalexp nFeature_originalexp cell_name  batch
# Cell1 SeuratProject              65963                  975     Cell1 Batch1
# Cell2 SeuratProject               1167                  301     Cell2 Batch1
# Cell3 SeuratProject              45665                  958     Cell3 Batch1
# Cell4 SeuratProject              26395                  911     Cell4 Batch1
# Cell5 SeuratProject               3059                  488     Cell5 Batch1
# Cell6 SeuratProject               1998                  393     Cell6 Batch1
#        group
# Cell1 Group2
# Cell2 Group2
# Cell3 Group3
# Cell4 Group2
# Cell5 Group3
# Cell6 Group3</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- seurat_result@assays[[&quot;originalexp&quot;]]@meta.features
head(gene_info)
#       gene_name de_gene BatchFacBatch1 BatchFacBatch2 DEFacGroup1 DEFacGroup2
# Gene1     Gene1      no      0.9098860      1.1054169           1           1
# Gene2     Gene2      no      0.9774161      0.9501320           1           1
# Gene3     Gene3      no      1.0541276      0.7597880           1           1
# Gene4     Gene4      no      1.1708139      0.7762219           1           1
# Gene5     Gene5      no      1.0017116      0.9963063           1           1
# Gene6     Gene6      no      0.7654326      1.1497335           1           1
#       DEFacGroup3
# Gene1           1
# Gene2           1
# Gene3           1
# Gene4           1
# Gene5           1
# Gene6           1</code></pre>
</div>
<div id="h5ad" class="section level4">
<h4>h5ad</h4>
<p>If we select <code>h5ad</code> format, it is not possible to return the result in R, so you can get the path where the <code>h5ad</code> files save to and we can go to the path and read it in <strong>Python</strong> by <code>scanpy.read_h5ad</code> function (if you have already installed Python and <strong>scanpy</strong> module).</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Splat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;h5ad&quot;,
  other_prior = list(batchCells = c(100, 100),
                     nGenes = 1000,
                     de.prob = 0.1,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 200
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 2
# Creating h5Seurat file for version 3.1.5.9900
# Adding counts for originalexp
# Adding data for originalexp
# No variable features found for originalexp
# Adding feature-level metadata for originalexp
# Validating h5Seurat file
# Adding data from originalexp as X
# Transfering meta.features to var
# Adding counts from originalexp as raw
# Transfering meta.features to raw/var
# Transfering meta.data to obs
# Your data has been save to /var/folders/1l/xmc98tgx0m37wxtbtwnl6h7c0000gn/T//RtmpY7iH84/20220914214036.h5ad
save_path &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;save_path&quot;]]
save_path
# [1] &quot;/var/folders/1l/xmc98tgx0m37wxtbtwnl6h7c0000gn/T//RtmpY7iH84/20220914214036.h5ad&quot;</code></pre>
<p>Now, we can go to the path and check the data. Here, we read the <code>h5ad</code> file in R using <strong>reticulate</strong> R package (note that <strong>Python</strong> and <strong>scanpy</strong> module must have been installed).</p>
<pre class="r"><code>## install.packages(&quot;reticulate&quot;)
scanpy &lt;- reticulate::import(&quot;scanpy&quot;)
data &lt;- scanpy$read_h5ad(save_path)
data ## Read h5ad file successfully
# AnnData object with n_obs × n_vars = 200 × 1000
#     obs: &#39;orig.ident&#39;, &#39;nCount_originalexp&#39;, &#39;nFeature_originalexp&#39;, &#39;cell_name&#39;, &#39;batch&#39;, &#39;group&#39;
#     var: &#39;gene_name&#39;, &#39;de_gene&#39;, &#39;BatchFacBatch1&#39;, &#39;BatchFacBatch2&#39;, &#39;DEFacGroup1&#39;, &#39;DEFacGroup2&#39;, &#39;DEFacGroup3&#39;</code></pre>
</div>
</div>
</div>
