---
title: "scDesign3"
output:
  blogdown::html_page:
    toc: yes
weight: 13
---

<!-- github markdown built using 
rmarkdown::render("vignettes/Splat.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```
Here scDesign3 method will be demonstrated clearly and hope that this document can help you.

## Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
```

### Default estimation

```{r, warning=FALSE}
estimate_result <- simmethods::scDesign3_estimation(
  ref_data = ref_data,
  verbose = TRUE,
  seed = 111
)
```


### Information of cell groups

If the information of cell groups is available, you can use another way to estimate the parameters.

```{r}
## cell groups
group_condition <- as.numeric(simmethods::group_condition)
```


```{r estimate, warning=FALSE}
estimate_result <- simmethods::scDesign3_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = group_condition),
  verbose = TRUE,
  seed = 111
)
```


## Simulating datasets using scDesign3

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Simulate two or more groups
3. Simulate two or more batches
4. Simulate data with cellular differentiation trajectory
5. Simulate spatial transcriptome data

### Datasets with default parameters

The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data.

```{r default_params}
simulate_result <- simmethods::scDesign3_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
```

{{% notice note %}}
We will get two or groups if information of cell groups is used in estimation step.
{{% /notice %}}

```{r}
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
```

```{r}
table(colData(SCE_result)$group)
```

{{% notice note %}}
We can not set the cell or gene number in scDesign3 unless a data frame which contains covariates of targeted simulated data from construct_data function is provided through new_covariate parameter.
{{% /notice %}}


### Simulate two or more groups

In scDesign3, we can not set `nGroups` directly and should specify the group labels of cells in the **estimation** step in `scDesign3_estimation` function.

We randomly assign four group labels to cells.
```{r estimate_random, warning=FALSE}
set.seed(666)
estimate_result <- simmethods::scDesign3_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = sample(1:4, ncol(ref_data), replace = TRUE)),
  verbose = TRUE,
  seed = 111
)
```


```{r different groups}
simulate_result <- simmethods::scDesign3_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
```

Check cell groups
```{r}
SCE_result <- simulate_result[["simulate_result"]]
table(colData(SCE_result)$group)
```


### Simulate two or more batches

In scDesign3, we can not set `nBatches` directly and should the batch labels of cells in the **estimation** step in `scDesign3_estimation` function.

{{% notice warning %}}
If you custom the simulated cell number which is not equal to that of the real data, the batch information for simulated cells is not returned.
{{% /notice %}}


### Simulate datasets with cellular differentiation trajectory

Before modeling from the reference data, we should construct `dynwrap::expression` data using `dynwrap::wrap_expression` function and perform trajectory inference using Slingshot using `tislingshot::ti_slingshot` or other methods.

```{r download slingshot}
if(!requireNamespace("tislingshot", quietly = TRUE)){
  devtools::install_github("dynverse/ti_slingshot/package/")
}
```


```{r dynwrap_expression}
dyn_data <- dynwrap::wrap_expression(
  counts = t(ref_data),
  expression = log2(t(ref_data) + 1)
)
dyn_data <- dynwrap::add_grouping(dyn_data, group_condition)
dyn_model <- dynwrap::infer_trajectory(dyn_data, tislingshot::ti_slingshot())
```


The `traj` parameter must be `TURE` and `dynwrap_data` is needed when you want to estimate parameters from trajectory data.
```{r est_traj, warning=FALSE}
estimate_result <- simmethods::scDesign3_estimation(
  ref_data = ref_data,
  other_prior = list(traj = TRUE,
                     group.condition = group_condition,
                     dynwrap_data = dyn_model),
  verbose = TRUE,
  seed = 111
)
```


Next, we can simulate trajectory data using scDesign3.
```{r sim_traj, warning=FALSE}
simulate_result <- simmethods::scDesign3_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  seed = 111
)
```


Before visualization, make sure that you have already installed several R packages:
```{r}
if(!requireNamespace("dyndimred", quietly = TRUE)){install.packages("dyndimred")}
if(!requireNamespace("dynplot", quietly = TRUE)){install.packages("dynplot")}
```

First we should wrap the data into a standard object:

```{r}
dyn_object <- dynwrap::wrap_expression(counts = t(simulate_result$simulate_result$count_data),
                                       expression = log2(t(simulate_result$simulate_result$count_data) + 1))
```

Next, we infer the trajectory using SlingShot which has been proved to be the most best method to do this:

```{r}
model <- dynwrap::infer_trajectory(dataset = dyn_object,
                                   method = tislingshot::ti_slingshot(),
                                   parameters = NULL,
                                   give_priors = NULL,
                                   seed = 666,
                                   verbose = TRUE)
```

Finally, we can plot the trajectory after performing dimensionality reduction:
```{r, fig.align='center', fig.height=6, fig.width=12}
dimred <- dyndimred::dimred_umap(dyn_object$expression)
dynplot::plot_dimred(model, dimred = dimred)
```

For more details about trajectory inference and visualization, please check [dynverse](https://dynverse.org/).


### Simulate spatial transcriptome data

scDesign3 can also simulate spatial transcriptome data. Besides the gene expression profile, users should also provide the spatial coordinates of each cell (spot). The reference data can be downloaded [here](https://zenodo.org/record/8251596/files/data118_spatial_OV.rds?download=1).

```{r load_sptial_data, message=FALSE, warning=FALSE, error=FALSE}
# Load data (downloaded from https://zenodo.org/record/8251596/files/data118_spatial_OV.rds?download=1)
data <- readRDS("../../../../preprocessed_data/data118_spatial_OV.rds")
ref_data <- t(as.matrix(data$data$counts))
```

In addition, we can set the spatial coordinates by `spatial.x` and `spatial.y` parameters.

```{r coordinate}
other_prior <- list(spatial.x = data$data_info$spatial_coordinate$x,
                    spatial.y = data$data_info$spatial_coordinate$y,
                    group.condition = as.numeric(as.factor(data$data_info$cluster_info)))
```

Execute the parameter estimation:
```{r, warning=FALSE}
estimate_result <- simmethods::scDesign3_estimation(
  ref_data = ref_data,
  other_prior = other_prior,
  verbose = T,
  seed = 10
)
```

Simulate spatial transcriptome data:
```{r simulate ST data}
simulate_result <- simmethods::scDesign3_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  seed = 111
)
```

Notably, the simulated cells can match the real ones, so that the spatial coordinates are the same as the previously input ones.

Visualize the spatial spots:
```{r plot_spatial, fig.align='center', fig.height=5, fig.width=6}
library(ggplot2)
location <- simulate_result$simulate_result$col_meta
p <- ggplot(location, aes(x = x, y = y))+
  geom_point(aes(color = group))+
  theme(panel.grid = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom")
p
```
