---
title: "zingeR"
output:
  blogdown::html_page:
    toc: yes
weight: 14
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-zinger" id="toc-simulating-datasets-using-zinger">Simulating datasets using zingeR</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
<li><a href="#simulate-two-groups" id="toc-simulate-two-groups">Simulate two groups</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/zingeR.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here zingeR method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<p>When you use zingeR to estimate parameters from a real dataset, you must input a numeric vector to specify the groups or plates that each cell comes from, like <code>other_prior = list(group.condition = the numeric vector)</code>.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data
group_condition &lt;- simmethods::group_condition
## group_condition can must be a numeric vector.
other_prior &lt;- list(group.condition = as.numeric(group_condition))</code></pre>
<p>Using <code>simmethods::zingeR_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::zingeR_estimation(ref_data = ref_data,
                                                 other_prior = other_prior,
                                                 verbose = T,
                                                 seed = 10)
# Estimating parameters using zingeR</code></pre>
</div>
<div id="simulating-datasets-using-zinger" class="section level2">
<h2>Simulating datasets using zingeR</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells and genes</li>
<li>Simulate two groups</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::zingeR_simulation(
  ref_data = ref_data,
  other_prior = other_prior,
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 2
# prob.group: 0.1
# fc.group: 2
# Loading required package: edgeR
# Loading required package: limma
# 
# Attaching package: &#39;limma&#39;
# The following object is masked from &#39;package:BiocGenerics&#39;:
# 
#     plotMA
# 
# Attaching package: &#39;edgeR&#39;
# The following object is masked from &#39;package:SingleCellExperiment&#39;:
# 
#     cpm
# Preparing dataset. Using existing parameters.
# Sampling.
# Calculating differential expression.
# Simulating data.</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 1 column
#         cell_name
#       &lt;character&gt;
# Cell1       Cell1
# Cell2       Cell2
# Cell3       Cell3
# Cell4       Cell4
# Cell5       Cell5
# Cell6       Cell6</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 3 columns
#         gene_name     de_gene     de_fc
#       &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
# Gene1       Gene1          no         0
# Gene2       Gene2          no         0
# Gene3       Gene3          no         0
# Gene4       Gene4          no         0
# Gene5       Gene5          no         0
# Gene6       Gene6          no         0</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>In zingeR, users can only set the number of cells and genes which is higher than the reference data.
Here, we simulate a new dataset with 1000 cells and 5000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::zingeR_simulation(
  ref_data = ref_data,
  other_prior = list(group.condition = as.numeric(group_condition),
                     nCells = 1000,
                     nGenes = 5000),
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  seed = 111
)
# nCells: 1000
# nGenes: 5000
# nGroups: 2
# prob.group: 0.1
# fc.group: 2
# Preparing dataset. Using existing parameters.
# Sampling.
# Calculating differential expression.
# Simulating data.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 5000 1000</code></pre>
</div>
<div id="simulate-two-groups" class="section level3">
<h3>Simulate two groups</h3>
<p>In zingeR, we can only simulate <strong>two groups</strong> and note that zingeR dose not return cell group information.</p>
<p>For demonstration, we will simulate two groups using the learned parameters. We can set <code>de.prob = 0.2</code> to simulate 20% genes as DEGs.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::zingeR_simulation(
  ref_data = ref_data,
  other_prior = list(group.condition = as.numeric(group_condition),
                     nCells = 1000,
                     nGenes = 5000,
                     de.prob = 0.2,
                     fc.group = 4),
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  seed = 111
)
# nCells: 1000
# nGenes: 5000
# nGroups: 2
# prob.group: 0.2
# fc.group: 4
# Preparing dataset. Using existing parameters.
# Sampling.
# Calculating differential expression.
# Simulating data.</code></pre>
<p>{{% notice note %}}
zingeR dose not return cell group information.
{{% /notice %}}</p>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 5000 1000</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
# yes 
# 0.2</code></pre>
</div>
</div>
