---
title: "zinbwave"
output:
  blogdown::html_page:
    toc: yes
weight: 13
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-zinbwave" id="toc-simulating-datasets-using-zinbwave">Simulating datasets using zinbwave</a></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/zinbwave.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here zinbwave method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data</code></pre>
<p>Using <code>simmethods::zinbwave_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::zinbwave_estimation(ref_data = ref_data,
                                                   verbose = T,
                                                   seed = 10)
# Estimating parameters using zinbwave
# Removing all zero genes...
# Fitting model...
# Create model:
# ok
# Initialize parameters:
# ok
# Optimize parameters:
# Iteration 1
# penalized log-likelihood = -1458528.51195789
# After dispersion optimization = -1842970.76424042
#    user  system elapsed 
#   6.818   0.562   7.599
# After right optimization = -1681437.25066416
# After orthogonalization = -1681437.25066416
#    user  system elapsed 
#   2.545   0.188   2.802
# After left optimization = -1615279.08884363
# After orthogonalization = -1615279.08884363
# Iteration 2
# penalized log-likelihood = -1615279.08884363
# After dispersion optimization = -1615279.08884363
#    user  system elapsed 
#   5.128   0.413   5.608
# After right optimization = -1613680.86211755
# After orthogonalization = -1613680.86211755
#    user  system elapsed 
#   2.016   0.158   2.199
# After left optimization = -1613424.05613962
# After orthogonalization = -1613424.05613962
# Iteration 3
# penalized log-likelihood = -1613424.05613962
# After dispersion optimization = -1613424.05613962
#    user  system elapsed 
#   4.531   0.374   4.962
# After right optimization = -1613329.33593763
# After orthogonalization = -1613329.33593763
#    user  system elapsed 
#   1.262   0.077   1.348
# After left optimization = -1613299.28560498
# After orthogonalization = -1613299.28560498
# Iteration 4
# penalized log-likelihood = -1613299.28560498
# ok</code></pre>
</div>
<div id="simulating-datasets-using-zinbwave" class="section level2">
<h2>Simulating datasets using zinbwave</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters.</p>
<p>The reference data contains 160 cells and 4000 genes, we can only simulate datasets with default parameters in zinbwave and then we will obtain a new data which has the same size as the reference data.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::zinbwave_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 1 column
#              Cell
#       &lt;character&gt;
# Cell1       Cell1
# Cell2       Cell2
# Cell3       Cell3
# Cell4       Cell4
# Cell5       Cell5
# Cell6       Cell6</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 1 column
#              Gene
#       &lt;character&gt;
# Gene1       Gene1
# Gene2       Gene2
# Gene3       Gene3
# Gene4       Gene4
# Gene5       Gene5
# Gene6       Gene6</code></pre>
</div>
