---
title: "zinbwave"
output:
  blogdown::html_page:
    toc: yes
weight: 13
---

<!-- github markdown built using 
rmarkdown::render("vignettes/zinbwave.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```
Here zinbwave method will be demonstrated clearly and hope that this document can help you.

## Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
```

Using `simmethods::zinbwave_estimation` command to execute the estimation step.

```{r estimate}
estimate_result <- simmethods::zinbwave_estimation(ref_data = ref_data,
                                                   verbose = T,
                                                   seed = 10)
```

## Simulating datasets using zinbwave

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters.

The reference data contains 160 cells and 4000 genes, we can only simulate datasets with default parameters in zinbwave and then we will obtain a new data which has the same size as the reference data.

```{r default_params}
simulate_result <- simmethods::zinbwave_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
```

```{r}
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
```

```{r}
head(colData(SCE_result))
```

```{r}
head(rowData(SCE_result))
```

