---
title: "SPARSim"
output:
  blogdown::html_page:
    toc: yes
weight: 18
---

<!-- github markdown built using 
rmarkdown::render("vignettes/SPARSim.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```
Here SPARSim method will be demonstrated clearly and hope that this document can help you.

## Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.


```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
```

### Estimation without cell group information
In SPARSim, users can estimate parameters from real data only using gene expression matrix.

Using `simmethods::SPARSim_estimation` command to execute the estimation step.

```{r, warning=FALSE}
estimate_result <- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = NULL,
  verbose = TRUE,
  seed = 111
)
```

### Estimation with cell group information
In addition, you can also input a numeric vector to specify the groups or plates that each cell comes from, like `other_prior = list(group.condition = the numeric vector)`.

```{r, warning=FALSE}
group <- as.numeric(simmethods::group_condition)
```

```{r, warning=FALSE}
estimate_result <- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = group),
  verbose = TRUE,
  seed = 111
)
```

### Estimation with spike-in genes
In SPARSim, ERCC spike-in genes can be used to estimate data parameters from the real data. In this case, the gene matrix must contain spike-in gene counts and hold the right ERCC spike-in gene names. Note that users must also input the dilution factor and volume (microliter) which the experiment used.

```{r, warning=FALSE}
group <- as.numeric(simmethods::group_condition)
other_prior <- list(group.condition = group,
                    dilution.factor = 50000,
                    volume = 0.01)
```

```{r, warning=FALSE}
estimate_result <- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = other_prior,
  verbose = TRUE,
  seed = 111
)
```

Check spike-in parameters

```{r}
spikein_params <- estimate_result[["estimate_result"]][["SPARSim_spikein_parameter"]]
str(spikein_params)
```


## Simulating datasets using SPARSim

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Simulate groups
3. Simulate batches

### Datasets with default parameters

The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data.

{{% notice note %}}
The simulated dataset will have one group of cells as no group information is used in estimation step.
{{% /notice %}}

```{r}
estimate_result <- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = NULL,
  verbose = TRUE,
  seed = 111
)
```

```{r default_params}
simulate_result <- simmethods::SPARSim_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
```

```{r}
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
```

```{r}
head(colData(SCE_result))
```

```{r}
head(rowData(SCE_result))
```

### Simulate groups

The number of groups simulated by SPARSim is determined by the group information used in estimation step. If the group information of two cell states is provided, then the simulated dataset will contain two groups. SPARSim also provides other parameters related to DEGs such as the proportion of DEGs (`de.prob`) and the fold change of DGEs (`fc.group`).

Estimating parameters using group information
```{r}
group <- as.numeric(simmethods::group_condition)
estimate_result <- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = group),
  verbose = TRUE,
  seed = 111
)
```


For demonstration, we will simulate two groups using the learned parameters. We can set `de.prob = 0.2` to simulate 20% genes as DEGs.
```{r}
simulate_result <- simmethods::SPARSim_simulation(
  other_prior = list(de.prob = 0.2,
                     fc.group = 4),
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  seed = 111
)
```


```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```

```{r}
## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
```


### Simulate batch

Users can simulate batches when `batch.condition` parameter is activated and just input the numeric vectors that specify the batch labels of cells.

For demonstration, we will simulate three batches using the learned parameters.
```{r}
simulate_result <- simmethods::SPARSim_simulation(
  other_prior = list(de.prob = 0.2,
                     fc.group = 4,
                     batch.condition = sample(1:3, 160, replace = TRUE)),
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  seed = 111
)
```


```{r}
## cell information
col_data <- simulate_result[["simulate_result"]][["col_meta"]]
```

```{r}
table(col_data$group)
```

```{r}
table(col_data$batch)
```
