---
title: "SPARSim"
output:
  blogdown::html_page:
    toc: yes
weight: 19
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a>
<ul>
<li><a href="#estimation-without-cell-group-information" id="toc-estimation-without-cell-group-information">Estimation without cell group information</a></li>
<li><a href="#estimation-with-cell-group-information" id="toc-estimation-with-cell-group-information">Estimation with cell group information</a></li>
<li><a href="#estimation-with-spike-in-genes" id="toc-estimation-with-spike-in-genes">Estimation with spike-in genes</a></li>
</ul></li>
<li><a href="#simulating-datasets-using-sparsim" id="toc-simulating-datasets-using-sparsim">Simulating datasets using SPARSim</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#simulate-groups" id="toc-simulate-groups">Simulate groups</a></li>
<li><a href="#simulate-batch" id="toc-simulate-batch">Simulate batch</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/SPARSim.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here SPARSim method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data</code></pre>
<div id="estimation-without-cell-group-information" class="section level3">
<h3>Estimation without cell group information</h3>
<p>In SPARSim, users can estimate parameters from real data only using gene expression matrix.</p>
<p>Using <code>simmethods::SPARSim_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = NULL,
  verbose = TRUE,
  seed = 111
)
# Estimating parameters using SPARSim
# [1] &quot;Experimental condition 1&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;</code></pre>
</div>
<div id="estimation-with-cell-group-information" class="section level3">
<h3>Estimation with cell group information</h3>
<p>In addition, you can also input a numeric vector to specify the groups or plates that each cell comes from, like <code>other_prior = list(group.condition = the numeric vector)</code>.</p>
<pre class="r"><code>group &lt;- as.numeric(simmethods::group_condition)</code></pre>
<pre class="r"><code>estimate_result &lt;- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = group),
  verbose = TRUE,
  seed = 111
)
# Estimating parameters using SPARSim
# [1] &quot;Experimental condition 1&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;
# [1] &quot;Experimental condition 2&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;</code></pre>
</div>
<div id="estimation-with-spike-in-genes" class="section level3">
<h3>Estimation with spike-in genes</h3>
<p>In SPARSim, ERCC spike-in genes can be used to estimate data parameters from the real data. In this case, the gene matrix must contain spike-in gene counts and hold the right ERCC spike-in gene names. Note that users must also input the dilution factor and volume (microliter) which the experiment used.</p>
<pre class="r"><code>group &lt;- as.numeric(simmethods::group_condition)
other_prior &lt;- list(group.condition = group,
                    dilution.factor = 50000,
                    volume = 0.01)</code></pre>
<pre class="r"><code>estimate_result &lt;- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = other_prior,
  verbose = TRUE,
  seed = 111
)
# Estimating parameters using SPARSim
# [1] &quot;Experimental condition 1&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;
# [1] &quot;Experimental condition 2&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;</code></pre>
<p>Check spike-in parameters</p>
<pre class="r"><code>spikein_params &lt;- estimate_result[[&quot;estimate_result&quot;]][[&quot;SPARSim_spikein_parameter&quot;]]
str(spikein_params)
# List of 4
#  $ spikein_set       :List of 1
#   ..$ spikein:List of 4
#   .. ..$ mix_name   : chr &quot;spikein&quot;
#   .. ..$ abundance  : Named num [1:50] 3613.2 903.3 225.8 112.9 56.5 ...
#   .. .. ..- attr(*, &quot;names&quot;)= chr [1:50] &quot;spikein_1&quot; &quot;spikein_2&quot; &quot;spikein_3&quot; &quot;spikein_4&quot; ...
#   .. ..$ variability: Named num [1:50] 0 0 0 0 0 0 0 0 0 0 ...
#   .. .. ..- attr(*, &quot;names&quot;)= chr [1:50] &quot;spikein_1&quot; &quot;spikein_2&quot; &quot;spikein_3&quot; &quot;spikein_4&quot; ...
#   .. ..$ ids        : chr [1:50] &quot;spikein_1&quot; &quot;spikein_2&quot; &quot;spikein_3&quot; &quot;spikein_4&quot; ...
#  $ spikein_sample    : chr [1:160] &quot;spikein&quot; &quot;spikein&quot; &quot;spikein&quot; &quot;spikein&quot; ...
#  $ spikein_proportion: num 0.05
#  $ spikein_ids       : chr [1:50] &quot;spikein_1&quot; &quot;spikein_2&quot; &quot;spikein_3&quot; &quot;spikein_4&quot; ...</code></pre>
</div>
</div>
<div id="simulating-datasets-using-sparsim" class="section level2">
<h2>Simulating datasets using SPARSim</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Simulate groups</li>
<li>Simulate batches</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data.</p>
<p>{{% notice note %}}
The simulated dataset will have one group of cells as no group information is used in estimation step.
{{% /notice %}}</p>
<pre class="r"><code>estimate_result &lt;- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = NULL,
  verbose = TRUE,
  seed = 111
)
# Estimating parameters using SPARSim
# [1] &quot;Experimental condition 1&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;</code></pre>
<pre class="r"><code>simulate_result &lt;- simmethods::SPARSim_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 1
# fc.group: 0
# de.prob: 0
# nBatches: 0
# Number of experimental conditions:  1 
# Number of genes:  4000 
# Number of cells:  160 
# Setting gene expression intensity...  
# Setting gene expression variability ...  
# Simulating biological variability ...  
# Simulating technical variability ...</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 2 columns
#         cell_name       group
#       &lt;character&gt; &lt;character&gt;
# Cell1       Cell1      Group1
# Cell2       Cell2      Group1
# Cell3       Cell3      Group1
# Cell4       Cell4      Group1
# Cell5       Cell5      Group1
# Cell6       Cell6      Group1</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 1 column
#         gene_name
#       &lt;character&gt;
# Gene1       Gene1
# Gene2       Gene2
# Gene3       Gene3
# Gene4       Gene4
# Gene5       Gene5
# Gene6       Gene6</code></pre>
</div>
<div id="simulate-groups" class="section level3">
<h3>Simulate groups</h3>
<p>The number of groups simulated by SPARSim is determined by the group information used in estimation step. If the group information of two cell states is provided, then the simulated dataset will contain two groups. SPARSim also provides other parameters related to DEGs such as the proportion of DEGs (<code>de.prob</code>) and the fold change of DGEs (<code>fc.group</code>).</p>
<p>Estimating parameters using group information</p>
<pre class="r"><code>group &lt;- as.numeric(simmethods::group_condition)
estimate_result &lt;- simmethods::SPARSim_estimation(
  ref_data = ref_data,
  other_prior = list(group.condition = group),
  verbose = TRUE,
  seed = 111
)
# Estimating parameters using SPARSim
# [1] &quot;Experimental condition 1&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;
# [1] &quot;Experimental condition 2&quot;
# [1] &quot;...estimating gene intensity&quot;
# [1] &quot;...estimating gene variability&quot;
# [1] &quot;...estimating library size&quot;
# [1] &quot;...creating SPARSim simulation parameter&quot;</code></pre>
<p>For demonstration, we will simulate two groups using the learned parameters. We can set <code>de.prob = 0.2</code> to simulate 20% genes as DEGs.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SPARSim_simulation(
  other_prior = list(de.prob = 0.2,
                     fc.group = 4),
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 2
# fc.group: 4
# de.prob: 0.2
# nBatches: 0
# Number of experimental conditions:  2 
# Number of genes:  4000 
# Number of cells:  160 
# Setting gene expression intensity...  
# Setting gene expression variability ...  
# Simulating biological variability ...  
# Simulating technical variability ...</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 4000  160</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
# yes 
# 0.2</code></pre>
</div>
<div id="simulate-batch" class="section level3">
<h3>Simulate batch</h3>
<p>Users can simulate batches when <code>batch.condition</code> parameter is activated and just input the numeric vectors that specify the batch labels of cells.</p>
<p>For demonstration, we will simulate three batches using the learned parameters.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SPARSim_simulation(
  other_prior = list(de.prob = 0.2,
                     fc.group = 4,
                     batch.condition = sample(1:3, 160, replace = TRUE)),
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 2
# fc.group: 4
# de.prob: 0.2
# nBatches: 3
# Number of experimental conditions:  2 
# Number of genes:  4000 
# Number of cells:  160 
# Setting gene expression intensity...  
# Setting gene expression variability ...  
# Simulating batch effects ...  
# Simulating biological variability ...  
# Simulating technical variability ...</code></pre>
<pre class="r"><code>## cell information
col_data &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]</code></pre>
<pre class="r"><code>table(col_data$group)
# 
# Group1 Group2 
#     80     80</code></pre>
<pre class="r"><code>table(col_data$batch)
# 
# Batch1 Batch2 Batch3 
#     63     43     54</code></pre>
</div>
</div>
