---
title: "Lun2"
output:
  blogdown::html_page:
    toc: yes
weight: 6
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-lun2" id="toc-simulating-datasets-using-lun2">Simulating datasets using Lun2</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
<li><a href="#simulate-two-or-more-groups" id="toc-simulate-two-or-more-groups">Simulate two or more groups</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/Lun2.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here Lun2 method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<p>When you use Lun2 to estimate parameters from a real dataset, you must input a numeric vector to specify the groups or plates that each cell comes from, like <code>other_prior = list(group.condition = the numeric vector)</code>.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data
group_condition &lt;- simmethods::group_condition
## group_condition can must be a numeric vector.
other_prior &lt;- list(group.condition = as.numeric(group_condition))</code></pre>
<p>Using <code>simmethods::Lun2_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::Lun2_estimation(ref_data = ref_data,
                                               other_prior = other_prior,
                                               verbose = T,
                                               seed = 10)
# Estimating parameters using Lun2
# Estimating number of groups...
# Warning in regularize.values(x, y, ties, missing(ties), na.rm = na.rm):
# collapsing to unique &#39;x&#39; values
# Computing normalisation factors...
# Warning in (function (x, sizes, min.mean = NULL, positive = FALSE, scaling =
# NULL) : encountered non-positive size factor estimates
# Estimating dispersions...
# Estimating gene means...
# Estimating plate effects...
# Estimating zero-inflated parameters...
# Warning in sqrt(diag(vc)[np]): NaNs produced
# Warning in sqrt(diag(vc)[np]): NaNs produced
# Warning in value[[3L]](cond): system is computationally singular: reciprocal
# condition number = 9.2681e-48FALSE
# Warning in sqrt(diag(vc)[np]): NaNs produced

# Warning in sqrt(diag(vc)[np]): NaNs produced

# Warning in sqrt(diag(vc)[np]): NaNs produced

# Warning in sqrt(diag(vc)[np]): NaNs produced

# Warning in sqrt(diag(vc)[np]): NaNs produced

# Warning in sqrt(diag(vc)[np]): NaNs produced
# Warning in value[[3L]](cond): system is computationally singular: reciprocal
# condition number = 4.1192e-48FALSE</code></pre>
</div>
<div id="simulating-datasets-using-lun2" class="section level2">
<h2>Simulating datasets using Lun2</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells and genes</li>
<li>Simulate two or more groups</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Lun2_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 3973
# nPlates: 2
# de.prob: 0
# fc.group: 3</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 3973  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 2 columns
#         cell_name    plate
#       &lt;character&gt; &lt;factor&gt;
# Cell1       Cell1        1
# Cell2       Cell2        1
# Cell3       Cell3        1
# Cell4       Cell4        1
# Cell5       Cell5        1
# Cell6       Cell6        1</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 1 column
#         gene_name
#       &lt;character&gt;
# Gene1       Gene1
# Gene2       Gene2
# Gene3       Gene3
# Gene4       Gene4
# Gene5       Gene5
# Gene6       Gene6</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>In Lun2, we can not set <code>nCells</code> directly and should set <code>cell.plates</code> instead. For example, if we want to simulate 500 cells, we can type <code>other_prior = list(cell.plates = rep(1, 500))</code>. For genes, we can just set <code>nGenes</code>.
Here, we simulate a new dataset with 500 cells and 2000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Lun2_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(cell.plates = rep(1, 500),
                     nGenes = 2000),
  seed = 111
)
# nCells: 500
# nGenes: 2000
# nPlates: 1
# de.prob: 0
# fc.group: 3
# Warning in splatter::lun2Simulate(parameters, verbose = verbose, zinb = zinb):
# Number of lib.sizes not equal to nCells. lib.sizes will be sampled.
# Warning in splatter::lun2Simulate(parameters, verbose = verbose, zinb = zinb):
# Number of gene parameters does not equal nGenes. Gene parameters will be
# sampled.</code></pre>
<p>{{% notice note %}}
The cell.plates parameter represents the sampling source of cells in real experiments.
{{% /notice %}}</p>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000  500</code></pre>
</div>
<div id="simulate-two-or-more-groups" class="section level3">
<h3>Simulate two or more groups</h3>
<p>In Lun2, we can not set <code>nGroups</code> directly and should set <code>cell.plates</code> instead. For example, if we want to simulate 2 groups, we can type <code>other_prior = list(cell.plates  = sample(1:2, 500, replace = TRUE))</code>. Note that the length of <code>cell.plates</code> numeric vector must be equal to the cell number.</p>
<p>For demonstration, we will simulate three groups using the learned parameters. We can set <code>de.prob = 0.2</code> to simulate 20% genes as DEGs.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Lun2_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(cell.plates = sample(1:2, 500, replace = TRUE),
                     nGenes = 2000,
                     de.prob = 0.2,
                     fc.group = 2),
  seed = 111
)
# nCells: 500
# nGenes: 2000
# nPlates: 2
# de.prob: 0.2
# fc.group: 2
# Warning in splatter::lun2Simulate(parameters, verbose = verbose, zinb = zinb):
# Number of lib.sizes not equal to nCells. lib.sizes will be sampled.
# Warning in splatter::lun2Simulate(parameters, verbose = verbose, zinb = zinb):
# Number of gene parameters does not equal nGenes. Gene parameters will be
# sampled.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$plate)
# 
#   1   2 
# 236 264</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
# yes 
# 0.2</code></pre>
</div>
</div>
