---
title: "SCRIP"
output:
  blogdown::html_page:
    toc: yes
weight: 7
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-scrip" id="toc-simulating-datasets-using-scrip">Simulating datasets using SCRIP</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#different-modes-in-scrip" id="toc-different-modes-in-scrip">Different modes in SCRIP</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
<li><a href="#simulate-two-or-more-groups" id="toc-simulate-two-or-more-groups">Simulate two or more groups</a></li>
<li><a href="#simulate-two-or-more-batches" id="toc-simulate-two-or-more-batches">Simulate two or more batches</a></li>
<li><a href="#simulate-more-groups-and-batches-simutaniously" id="toc-simulate-more-groups-and-batches-simutaniously">Simulate more groups and batches simutaniously</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/SCRIP.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here SCRIP method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data
dim(ref_data)
# [1] 4000  160</code></pre>
<p>Using <code>simmethods::SCRIP_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::SCRIP_estimation(ref_data = ref_data,
                                                verbose = T,
                                                seed = 10)
# Estimating parameters using SCRIP</code></pre>
</div>
<div id="simulating-datasets-using-scrip" class="section level2">
<h2>Simulating datasets using SCRIP</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Different modes in SCRIP</li>
<li>Determin the number of cells and genes</li>
<li>Simulate two or more groups</li>
<li>Simulate two or more batches</li>
<li>Simulate more groups and batches simutaniously</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group and one batch of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  ref_data = ref_data,
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1</code></pre>
<p>{{% notice note %}}
The reference data must be input when simulating new datasets.
{{% /notice %}}</p>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 3 columns
#         cell_name       batch       group
#       &lt;character&gt; &lt;character&gt; &lt;character&gt;
# Cell1       Cell1      Batch1      Group1
# Cell2       Cell2      Batch1      Group1
# Cell3       Cell3      Batch1      Group1
# Cell4       Cell4      Batch1      Group1
# Cell5       Cell5      Batch1      Group1
# Cell6       Cell6      Batch1      Group1</code></pre>
<p>Time consuming:</p>
<pre class="r"><code>simulate_result$simulate_detection$Elapsed_Time_sec
# [1] 20.288</code></pre>
</div>
<div id="different-modes-in-scrip" class="section level3">
<h3>Different modes in SCRIP</h3>
<p>SCRIP contains five different simulation modes, and you can specify which mode do you use (default is GP-trendedBCV):</p>
<ol style="list-style-type: decimal">
<li>GP-trendedBCV</li>
<li>GP-commonBCV</li>
<li>BGP-commonBCV</li>
<li>BP</li>
<li>BGP-trendedBCV</li>
</ol>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(parameters = estimate_result[[&quot;estimate_result&quot;]],
                                                ref_data = ref_data,
                                                other_prior = list(mode = &quot;BP&quot;),
                                                return_format = &quot;list&quot;,
                                                verbose = TRUE,
                                                seed = 111)
# nCells: 160
# nGenes: 4000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1
# Simulating datasets using SCRIP</code></pre>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(parameters = estimate_result[[&quot;estimate_result&quot;]],
                                                ref_data = ref_data,
                                                other_prior = list(mode = &quot;BGP-commonBCV&quot;),
                                                return_format = &quot;list&quot;,
                                                verbose = TRUE,
                                                seed = 111)
# nCells: 160
# nGenes: 4000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1
# Simulating datasets using SCRIP</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>In SCRIP, we can not set <code>nCells</code> directly and should set <code>batchCells</code> instead. For example, if we want to simulate 500 cells, we can type <code>other_prior = list(batchCells = 500)</code>. For genes, we can just set <code>nGenes</code>.
Here, we simulate a new dataset with 500 cells and 1000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  ref_data = ref_data,
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = 500,
                     nGenes = 1000),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 1
# de.prob: 0.1
# nBatches: 1</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
</div>
<div id="simulate-two-or-more-groups" class="section level3">
<h3>Simulate two or more groups</h3>
<p>In SCRIP, we can not set <code>nGroups</code> directly and should set <code>prob.group</code> instead. For example, if we want to simulate 2 groups, we can type <code>other_prior = list(prob.group = c(0.5, 0.5))</code>. Note that the sum of <code>prob.group</code> numeric vector must equal to 1, so we can also set <code>prob.group = c(0.3, 0.7)</code>.</p>
<p>In addtion, if we want to simulate three or more groups, we should obey the rules:</p>
<ul>
<li>The length of <code>prob.group</code> vector must always equal to the number of groups.</li>
<li>The sum of <code>prob.group</code> numeric vector must equal to 1.</li>
</ul>
<p>For demonstration, we will simulate three groups using the learned parameters.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  ref_data = ref_data,
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = 500,
                     nGenes = 1000,
                     prob.group = c(0.1, 0.3, 0.6)),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 3
# de.prob: 0.1
# nBatches: 1</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$group)
# 
# Group1 Group2 Group3 
#     46    156    298</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.1
# yes 
# 0.1</code></pre>
<p>We can see that the proportion of differential expressed genes is 0.1 (equals to the default). Next, if we want to know the fold change between two groups, we will do division with the groups that we are interested in.</p>
<pre class="r"><code>## fc between group2 and group1
fc_group1_to_group2 &lt;- gene_info$DEFacGroup2/gene_info$DEFacGroup1
## fc between group3 and group1
fc_group1_to_group3 &lt;- gene_info$DEFacGroup3/gene_info$DEFacGroup1
## fc between group3 and group2
fc_group2_to_group3 &lt;- gene_info$DEFacGroup3/gene_info$DEFacGroup2</code></pre>
</div>
<div id="simulate-two-or-more-batches" class="section level3">
<h3>Simulate two or more batches</h3>
<p>In SCRIP, we can not set <code>nBatches</code> directly and should set <code>batchCells</code> instead. For example, if we want to simulate 2 batches, we can type <code>other_prior = list(batchCells = c(250, 250))</code>. Note that the sum of <code>batchCells</code> numeric vector represents the total number of cells and the length of the vector equals to the number of batches.</p>
<p>In addtion, if we want to simulate three or more batches, we should obey the rules:</p>
<ul>
<li>The length of <code>batchCells</code> vector always equals to the number of batches.</li>
<li>The sum of <code>batchCells</code> numeric vector represents the total number of cells.</li>
</ul>
<p>For demonstration, we will simulate three batches using the learned parameters.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  ref_data = ref_data,
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = c(200, 300),
                     nGenes = 1000),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 1
# de.prob: 0.1
# nBatches: 2</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#    200    300</code></pre>
</div>
<div id="simulate-more-groups-and-batches-simutaniously" class="section level3">
<h3>Simulate more groups and batches simutaniously</h3>
<p>As mentioned before, we can set <code>prob.group</code> and <code>batchCells</code> to determine the number of groups and batches and we can also set <code>de.prob</code> to specify the proportion of DEGs. Here, we simulate a dataset with following settings:</p>
<ul>
<li>1000 cells</li>
<li>2000 genes</li>
<li>three groups with 0.2 proportion of DEGs</li>
<li>two batches</li>
</ul>
<pre class="r"><code>simulate_result &lt;- simmethods::SCRIP_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  ref_data = ref_data,
  return_format = &quot;list&quot;,
  other_prior = list(batchCells = c(500, 500),
                     nGenes = 2000,
                     de.prob = 0.2,
                     prob.group = c(0.2, 0.3, 0.5)),
  seed = 111
)
# nCells: 1000
# nGenes: 2000
# nGroups: 3
# de.prob: 0.2
# nBatches: 2</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000 1000</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#    500    500
table(cell_info$group)
# 
# Group1 Group2 Group3 
#    186    321    493</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result)
#    yes 
# 0.1685</code></pre>
</div>
</div>
