---
title: "powsimR"
output:
  blogdown::html_page:
    toc: yes
weight: 8
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a>
<ul>
<li><a href="#estimate-parameters-without-ercc-spike-in" id="toc-estimate-parameters-without-ercc-spike-in">Estimate parameters without ERCC spike-in</a></li>
<li><a href="#estimate-parameters-with-ercc-spike-in" id="toc-estimate-parameters-with-ercc-spike-in">Estimate parameters with ERCC spike-in</a></li>
</ul></li>
<li><a href="#simulating-datasets-using-powsimr" id="toc-simulating-datasets-using-powsimr">Simulating datasets using powsimR</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
<li><a href="#simulate-two-or-more-groups" id="toc-simulate-two-or-more-groups">Simulate two or more groups</a></li>
<li><a href="#simulate-two-or-more-batches" id="toc-simulate-two-or-more-batches">Simulate two or more batches</a></li>
<li><a href="#simulate-more-groups-and-batches-simutaniously" id="toc-simulate-more-groups-and-batches-simutaniously">Simulate more groups and batches simutaniously</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/powsimR.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here powsimR method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data</code></pre>
<p>powsimR provides some choices for users to select suitable parameters according to different types of data, platforms, normalization methods, distributions and so on.</p>
<ol style="list-style-type: decimal">
<li><code>RNAseq</code> “bulk” or “singlecell” (default).</li>
<li><code>Protocol</code> Options are “UMI” (default) (e.g. 10X Genomics, CEL-seq2) or “Read” (e.g. Smart-seq2).</li>
<li><code>Distribution</code> “NB” (default) for negative binomial or “ZINB” for zero-inflated negative binomial distribution fitting.</li>
<li><code>Normalisation</code> “TMM” (default), “MR”, “PosCounts”, “UQ”, “scran”, “Linnorm”, “SCnorm”, “Census”, “depth”, “none”.</li>
</ol>
<div id="estimate-parameters-without-ercc-spike-in" class="section level3">
<h3>Estimate parameters without ERCC spike-in</h3>
<pre class="r"><code>estimate_result &lt;- powsimR_estimation(
  ref_data = ref_data,
  other_prior = list(RNAseq = &quot;singlecell&quot;,
                     Protocol = &quot;UMI&quot;,
                     Normalisation = &quot;scran&quot;),
  verbose = TRUE,
  seed = 111)
# Warning: replacing previous import &#39;DECENT::lrTest&#39; by &#39;MAST::lrTest&#39; when
# loading &#39;powsimR&#39;
# Warning: replacing previous import &#39;penalized::predict&#39; by &#39;stats::predict&#39; when
# loading &#39;powsimR&#39;
# Warning: replacing previous import &#39;zinbwave::glmWeightedF&#39; by
# &#39;zingeR::glmWeightedF&#39; when loading &#39;powsimR&#39;
# Estimating parameters using estimateParam function</code></pre>
</div>
<div id="estimate-parameters-with-ercc-spike-in" class="section level3">
<h3>Estimate parameters with ERCC spike-in</h3>
<p>powsimR also provides an another choice to estimate parameters (not neccessary) via spike-ins. If users want to use this, make sure that the reference data must contain ERCC spike-in counts. In addtion, users must set dilution.factor and volume information by <code>other_prior = list(dilution.factor = xxx, volume = xxx)</code>.</p>
<pre class="r"><code>rownames(ref_data)[grep(pattern = &quot;^ERCC&quot;, x = rownames(ref_data))]</code></pre>
<pre><code>#  [1] &quot;ERCC-00002&quot; &quot;ERCC-00003&quot; &quot;ERCC-00004&quot; &quot;ERCC-00009&quot; &quot;ERCC-00014&quot;
#  [6] &quot;ERCC-00019&quot; &quot;ERCC-00022&quot; &quot;ERCC-00025&quot; &quot;ERCC-00034&quot; &quot;ERCC-00035&quot;
# [11] &quot;ERCC-00042&quot; &quot;ERCC-00043&quot; &quot;ERCC-00044&quot; &quot;ERCC-00046&quot; &quot;ERCC-00051&quot;
# [16] &quot;ERCC-00053&quot; &quot;ERCC-00054&quot; &quot;ERCC-00059&quot; &quot;ERCC-00060&quot; &quot;ERCC-00062&quot;
# [21] &quot;ERCC-00069&quot; &quot;ERCC-00071&quot; &quot;ERCC-00074&quot; &quot;ERCC-00076&quot; &quot;ERCC-00078&quot;
# [26] &quot;ERCC-00079&quot; &quot;ERCC-00084&quot; &quot;ERCC-00092&quot; &quot;ERCC-00095&quot; &quot;ERCC-00096&quot;
# [31] &quot;ERCC-00099&quot; &quot;ERCC-00108&quot; &quot;ERCC-00111&quot; &quot;ERCC-00112&quot; &quot;ERCC-00113&quot;
# [36] &quot;ERCC-00116&quot; &quot;ERCC-00130&quot; &quot;ERCC-00131&quot; &quot;ERCC-00136&quot; &quot;ERCC-00144&quot;
# [41] &quot;ERCC-00145&quot; &quot;ERCC-00148&quot; &quot;ERCC-00154&quot; &quot;ERCC-00157&quot; &quot;ERCC-00160&quot;
# [46] &quot;ERCC-00162&quot; &quot;ERCC-00163&quot; &quot;ERCC-00165&quot; &quot;ERCC-00170&quot; &quot;ERCC-00171&quot;</code></pre>
<p>{{% notice note %}}
Make sure there are ERCC names in reference data and users must input the dilution.factor and volume (microliter) to determine the concentration of ERCC molecules.
{{% /notice %}}</p>
<pre class="r"><code>estimate_result &lt;- powsimR_estimation(
  ref_data = ref_data,
  other_prior = list(RNAseq = &quot;singlecell&quot;,
                     Protocol = &quot;UMI&quot;,
                     Normalisation = &quot;scran&quot;,
                     dilution.factor = 50000,
                     volume = 1),
  verbose = TRUE,
  seed = 111)
# Estimating parameters using estimateParam function</code></pre>
</div>
</div>
<div id="simulating-datasets-using-powsimr" class="section level2">
<h2>Simulating datasets using powsimR</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells and genes</li>
<li>Simulate two or more groups</li>
<li>Simulate two or more batches</li>
<li>Simulate more groups and batches simutaniously</li>
</ol>
<p>powsimR provides some choices for users to select suitable parameters according to different normalization methods, and methods for differential expressed analysis.</p>
<ol style="list-style-type: decimal">
<li><code>Normalisation</code> “TMM” (default), “MR”, “PosCounts”, “UQ”, “scran”, “Linnorm”, “sctransform”, “SCnorm”, “Census”, “depth”.</li>
<li><code>DEmethod</code> “T-Test”, “edgeR-LRT”, “edgeR-QL”, “edgeR-zingeR”, “edgeR-ZINB-WaVE”, “limma-voom”, “limma-trend” (default), “DESeq2”, “DESeq2-zingeR”, “DESeq2-ZINB-WaVE”, “ROTS”, “baySeq”, “NOISeq”, “EBSeq”, “MAST”, “BPSC”, “scDD”, “DECENT”.</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group and one batch of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::powsimR_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  verbose = TRUE,
  seed = 111)
# nCells: 160
# nGenes: 4000
# nGroups: 2
# de.prob: 0.1
# fc.group: 2
# nBatches: 1
# 
# Setup Seed: 111
# You have chosen to simulate the expression of 4000 genes, which will be randomly drawn with replacement from the observed expression of 4000 genes.
# Simulating datasets using powsimR
# limma-trend is developed for bulk RNA-seq experiments.
# Preparing output arrays.
# 
#   SIMULATION   NUMBER   1
# Generating gene expression.
# Generating spike-in expression.
# 80 vs. 80
# Applying TMM normalisation
# Applying limma-trend for DE analysis on raw count data.
# Saving raw simulated counts.
# Estimating moments of raw count data.</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 2 columns
#         cell_name       group
#       &lt;character&gt; &lt;character&gt;
# Cell1       Cell1      Group1
# Cell2       Cell2      Group1
# Cell3       Cell3      Group1
# Cell4       Cell4      Group1
# Cell5       Cell5      Group1
# Cell6       Cell6      Group1</code></pre>
<p>Time consuming:</p>
<pre class="r"><code>simulate_result$simulate_detection$Elapsed_Time_sec
# [1] 0.982</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>Here, we simulate a new dataset with 500 cells and 1000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::powsimR_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  other_prior = list(nCells = 500,
                     nGenes = 1000),
  return_format = &quot;list&quot;,
  verbose = TRUE,
  seed = 111)
# nCells: 500
# nGenes: 1000
# nGroups: 2
# de.prob: 0.1
# fc.group: 2
# nBatches: 1
# 
# Setup Seed: 111
# You have chosen to simulate the expression of 1000 genes, which will be randomly drawn without replacement from the observed expression of 4000 genes.
# Simulating datasets using powsimR
# limma-trend is developed for bulk RNA-seq experiments.
# Preparing output arrays.
# 
#   SIMULATION   NUMBER   1
# Generating gene expression.
# Generating spike-in expression.
# 250 vs. 250
# Applying TMM normalisation
# Applying limma-trend for DE analysis on raw count data.
# Saving raw simulated counts.
# Estimating moments of raw count data.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
</div>
<div id="simulate-two-or-more-groups" class="section level3">
<h3>Simulate two or more groups</h3>
<p>In powsimR, we can not set <code>nGroups</code> directly and should set <code>prob.group</code> instead. For example, if we want to simulate 2 groups, we can type <code>other_prior = list(prob.group = c(0.5, 0.5))</code>. Note that the sum of <code>prob.group</code> numeric vector must equal to 1, so we can also set <code>prob.group = c(0.3, 0.7)</code>.</p>
<p>In addtion, if we want to simulate three or more groups, we should obey the rules:</p>
<ul>
<li>The length of <code>prob.group</code> vector must always be <strong>2 when using powsinR</strong>.</li>
<li>The sum of <code>prob.group</code> numeric vector must equal to 1.</li>
</ul>
<p>For demonstration, we will simulate two groups using the learned parameters. (20% DEGs and 4 fold change)</p>
<pre class="r"><code>simulate_result &lt;- simmethods::powsimR_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(nCells = 500,
                     nGenes = 1000,
                     prob.group = c(0.3, 0.7),
                     de.prob = 0.2,
                     fc.group = 4),
  seed = 111
)
# nCells: 500
# nGenes: 1000
# nGroups: 2
# de.prob: 0.2
# fc.group: 4
# nBatches: 1
# 
# You have chosen to simulate the expression of 1000 genes, which will be randomly drawn without replacement from the observed expression of 4000 genes.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$group)
# 
# Group1 Group2 
#    150    350</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
# yes 
# 0.2</code></pre>
</div>
<div id="simulate-two-or-more-batches" class="section level3">
<h3>Simulate two or more batches</h3>
<p>In powsimR, we can not set <code>nBatches</code> directly and should set <code>prob.batch</code> instead. For example, if we want to simulate 2 batches, we can type <code>other_prior = list(prob.batch = c(0.5, 0.5))</code>. Note that the sum of <code>prob.batch</code> numeric vector represents the total number of cells and the length of the vector equals to the number of batches.</p>
<p>In addtion, if we want to simulate three or more batches, we should obey the rules:</p>
<ul>
<li>The length of <code>prob.batch</code> vector must always equal to the number of batches</li>
<li>The sum of <code>prob.batch</code> numeric vector must equal to 1.</li>
</ul>
<p>For demonstration, we will simulate two batches using the learned parameters. (2 fold change)</p>
<pre class="r"><code>simulate_result &lt;- simmethods::powsimR_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  other_prior = list(prob.batch = c(0.4, 0.6),
                     fc.batch = 2),
  return_format = &quot;list&quot;,
  verbose = TRUE,
  seed = 111)
# nCells: 160
# nGenes: 4000
# nGroups: 2
# de.prob: 0.1
# fc.group: 2
# nBatches: 2
# fc.batch: 2
# Setup Seed: 111
# You have chosen to simulate the expression of 4000 genes, which will be randomly drawn with replacement from the observed expression of 4000 genes.
# Simulating datasets using powsimR
# limma-trend is developed for bulk RNA-seq experiments.
# Preparing output arrays.
# 
#   SIMULATION   NUMBER   1
# Generating gene expression.
# Generating spike-in expression.
# 32 vs. 32
# Applying TMM normalisation
# Applying limma-trend for DE analysis on raw count data.
# Saving raw simulated counts.
# Estimating moments of raw count data.
# 48 vs. 48
# Applying TMM normalisation
# Applying limma-trend for DE analysis on raw count data.
# Saving raw simulated counts.
# Estimating moments of raw count data.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 4000  160</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#     64     96</code></pre>
</div>
<div id="simulate-more-groups-and-batches-simutaniously" class="section level3">
<h3>Simulate more groups and batches simutaniously</h3>
<p>As mentioned before, we can set <code>prob.group</code> and <code>prob.batch</code> to determine the number of groups and batches and we can also set <code>de.prob</code> to specify the proportion of DEGs. Here, we simulate a dataset with following settings:</p>
<ul>
<li>1000 cells</li>
<li>2000 genes</li>
<li>two groups with 0.2 proportion of DEGs</li>
<li>two batches</li>
</ul>
<pre class="r"><code>simulate_result &lt;- simmethods::powsimR_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(nCells = 1000,
                     nGenes = 2000,
                     de.prob = 0.2,
                     prob.group = c(0.4, 0.6),
                     prob.batch = c(0.5, 0.5)),
  seed = 111
)
# nCells: 1000
# nGenes: 2000
# nGroups: 2
# de.prob: 0.2
# fc.group: 2
# nBatches: 2
# fc.batch: 2
# You have chosen to simulate the expression of 2000 genes, which will be randomly drawn without replacement from the observed expression of 4000 genes.</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000 1000</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$batch)
# 
# Batch1 Batch2 
#    500    500
table(cell_info$group)
# 
# Group1 Group2 
#    400    600</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result)
# yes 
# 0.2</code></pre>
</div>
</div>
