---
title: "powsimR"
output:
  blogdown::html_page:
    toc: yes
weight: 8
---

<!-- github markdown built using 
rmarkdown::render("vignettes/powsimR.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```
Here powsimR method will be demonstrated clearly and hope that this document can help you.

## Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by `simmethods:data` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- simmethods::data
```

powsimR provides some choices for users to select suitable parameters according to different types of data, platforms, normalization methods, distributions and so on.

1. `RNAseq` "bulk" or "singlecell" (default).
2. `Protocol` Options are "UMI" (default) (e.g. 10X Genomics, CEL-seq2) or "Read" (e.g. Smart-seq2).
3. `Distribution` "NB" (default) for negative binomial or "ZINB" for zero-inflated negative binomial distribution fitting.
4. `Normalisation` "TMM" (default), "MR", "PosCounts", "UQ", "scran", "Linnorm", "SCnorm", "Census", "depth", "none".

### Estimate parameters without ERCC spike-in

```{r estimate without ERCC, message=FALSE}
estimate_result <- powsimR_estimation(
  ref_data = ref_data,
  other_prior = list(RNAseq = "singlecell",
                     Protocol = "UMI",
                     Normalisation = "scran"),
  verbose = TRUE,
  seed = 111)
```

### Estimate parameters with ERCC spike-in

powsimR also provides an another choice to estimate parameters (not neccessary) via spike-ins. If users want to use this, make sure that the reference data must contain ERCC spike-in counts. In addtion, users must set dilution.factor and volume information by `other_prior = list(dilution.factor = xxx, volume = xxx)`.

```{r ERCC names, collapse=FALSE, warning=FALSE}
rownames(ref_data)[grep(pattern = "^ERCC", x = rownames(ref_data))]
```


{{% notice note %}}
Make sure there are ERCC names in reference data and users must input the dilution.factor and volume (microliter) to determine the concentration of ERCC molecules.
{{% /notice %}}

```{r estimate with ERCC, message=FALSE, warning=FALSE}
estimate_result <- powsimR_estimation(
  ref_data = ref_data,
  other_prior = list(RNAseq = "singlecell",
                     Protocol = "UMI",
                     Normalisation = "scran",
                     dilution.factor = 50000,
                     volume = 1),
  verbose = TRUE,
  seed = 111)
```


## Simulating datasets using powsimR

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Determin the number of cells and genes
3. Simulate two or more groups
4. Simulate two or more batches
5. Simulate more groups and batches simutaniously

powsimR provides some choices for users to select suitable parameters according to different normalization methods, and methods for differential expressed analysis.

1. `Normalisation` "TMM" (default), "MR", "PosCounts", "UQ", "scran", "Linnorm", "sctransform", "SCnorm", "Census", "depth".
2. `DEmethod` "T-Test", "edgeR-LRT", "edgeR-QL", "edgeR-zingeR", "edgeR-ZINB-WaVE", "limma-voom", "limma-trend" (default), "DESeq2", "DESeq2-zingeR", "DESeq2-ZINB-WaVE", "ROTS", "baySeq", "NOISeq", "EBSeq", "MAST", "BPSC", "scDD", "DECENT".

### Datasets with default parameters

The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group and one batch of cells.

```{r default_params}
simulate_result <- simmethods::powsimR_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  verbose = TRUE,
  seed = 111)
```

```{r}
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
```

```{r}
head(colData(SCE_result))
```

Time consuming:
```{r}
simulate_result$simulate_detection$Elapsed_Time_sec
```

### Determin the number of cells and genes

Here, we simulate a new dataset with 500 cells and 1000 genes:
```{r customed cell and gene number}
simulate_result <- simmethods::powsimR_simulation(
  parameters = estimate_result[["estimate_result"]],
  other_prior = list(nCells = 500,
                     nGenes = 1000),
  return_format = "list",
  verbose = TRUE,
  seed = 111)
```

```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```


### Simulate two or more groups

In powsimR, we can not set `nGroups` directly and should set `prob.group` instead. For example, if we want to simulate 2 groups, we can type `other_prior = list(prob.group = c(0.5, 0.5))`. Note that the sum of `prob.group` numeric vector must equal to 1, so we can also set `prob.group = c(0.3, 0.7)`.

In addtion, if we want to simulate three or more groups, we should obey the rules:

* The length of `prob.group` vector must always be **2 when using powsinR**.
* The sum of `prob.group` numeric vector must equal to 1.

For demonstration, we will simulate two groups using the learned parameters. (20% DEGs and 4 fold change)
```{r groups}
simulate_result <- simmethods::powsimR_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(nCells = 500,
                     nGenes = 1000,
                     prob.group = c(0.3, 0.7),
                     de.prob = 0.2,
                     fc.group = 4),
  seed = 111
)
```

```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```


```{r}
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$group)
```

```{r}
## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
```


### Simulate two or more batches

In powsimR, we can not set `nBatches` directly and should set `prob.batch` instead. For example, if we want to simulate 2 batches, we can type `other_prior = list(prob.batch = c(0.5, 0.5))`. Note that the sum of `prob.batch` numeric vector represents the total number of cells and the length of the vector equals to the number of batches.

In addtion, if we want to simulate three or more batches, we should obey the rules:

* The length of `prob.batch` vector must always equal to the number of batches
* The sum of `prob.batch` numeric vector must equal to 1.

For demonstration, we will simulate two batches using the learned parameters. (2 fold change)
```{r batches, results='hold'}
simulate_result <- simmethods::powsimR_simulation(
  parameters = estimate_result[["estimate_result"]],
  other_prior = list(prob.batch = c(0.4, 0.6),
                     fc.batch = 2),
  return_format = "list",
  verbose = TRUE,
  seed = 111)
```

```{r, results='hold'}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```

```{r, results='hold'}
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$batch)
```


### Simulate more groups and batches simutaniously

As mentioned before, we can set `prob.group` and `prob.batch` to determine the number of groups and batches and we can also set `de.prob` to specify the proportion of DEGs. Here, we simulate a dataset with following settings:

* 1000 cells
* 2000 genes
* two groups with 0.2 proportion of DEGs
* two batches

```{r group and batch}
simulate_result <- simmethods::powsimR_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(nCells = 1000,
                     nGenes = 2000,
                     de.prob = 0.2,
                     prob.group = c(0.4, 0.6),
                     prob.batch = c(0.5, 0.5)),
  seed = 111
)
```

```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```

```{r}
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$batch)
table(cell_info$group)
```

```{r}
## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result)
```
