---
title: "POWSC"
output:
  blogdown::html_page:
    toc: yes
weight: 9
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-powsc" id="toc-simulating-datasets-using-powsc">Simulating datasets using POWSC</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells" id="toc-determin-the-number-of-cells">Determin the number of cells</a></li>
<li><a href="#simulate-two-groups" id="toc-simulate-two-groups">Simulate two groups</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/POWSC.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here POWSC method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data generated by <code>scater::mockSCE</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- counts(scater::mockSCE())
dim(ref_data)
# [1] 2000  200</code></pre>
<p>Using <code>simmethods::POWSC_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::POWSC_estimation(ref_data = ref_data,
                                                verbose = T,
                                                seed = 10)
# Estimating parameters using POWSC</code></pre>
</div>
<div id="simulating-datasets-using-powsc" class="section level2">
<h2>Simulating datasets using POWSC</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells</li>
<li>Simulate two groups</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 200 cells and 2000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group of cells.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::POWSC_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 200
# nGenes: 2000
# nGroups: 2
# de.prob: 0.1</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 2000  200</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 2 columns
#         cell_name       group
#       &lt;character&gt; &lt;character&gt;
# Cell1       Cell1      Group1
# Cell2       Cell2      Group1
# Cell3       Cell3      Group1
# Cell4       Cell4      Group1
# Cell5       Cell5      Group1
# Cell6       Cell6      Group1</code></pre>
<pre class="r"><code>head(rowData(SCE_result))
# DataFrame with 6 rows and 2 columns
#         gene_name     de_gene
#       &lt;character&gt; &lt;character&gt;
# Gene1       Gene1          no
# Gene2       Gene2          no
# Gene3       Gene3         yes
# Gene4       Gene4          no
# Gene5       Gene5         yes
# Gene6       Gene6          no</code></pre>
</div>
<div id="determin-the-number-of-cells" class="section level3">
<h3>Determin the number of cells</h3>
<p>In POWSC, we can set <code>nCells</code> directly. For example, if we want to simulate 500 cells, we can type <code>other_prior = list(nCells = 500)</code>.
Here, we simulate a new dataset with 500 cells:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::POWSC_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(nCells = 500),
  seed = 111
)
# nCells: 500
# nGenes: 2000
# nGroups: 2
# de.prob: 0.1</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000  500</code></pre>
</div>
<div id="simulate-two-groups" class="section level3">
<h3>Simulate two groups</h3>
<p>POWSC will automatically simulate two cell groups by default. Users can set <code>de.prob</code> to specify the proportion of DEGs between two groups.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::POWSC_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(nCells = 500,
                     de.prob = 0.2),
  seed = 111
)
# nCells: 500
# nGenes: 2000
# nGroups: 2
# de.prob: 0.2</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 2000  500</code></pre>
<pre class="r"><code>## cell information
cell_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;col_meta&quot;]]
table(cell_info$group)
# 
# Group1 Group2 
#    250    250</code></pre>
<pre class="r"><code>## gene information
gene_info &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;row_meta&quot;]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
#    yes 
# 0.1895</code></pre>
</div>
</div>
