---
title: "POWSC"
output:
  blogdown::html_page:
    toc: yes
weight: 9
---

<!-- github markdown built using 
rmarkdown::render("vignettes/POWSC.Rmd", output_format = rmarkdown::github_document())
-->

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```
Here POWSC method will be demonstrated clearly and hope that this document can help you.

## Estimating parameters from a real dataset

Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data generated by `scater::mockSCE` command.

```{r load_data, message=FALSE, warning=FALSE, error=FALSE}
library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data <- counts(scater::mockSCE())
dim(ref_data)
```

Using `simmethods::POWSC_estimation` command to execute the estimation step.

```{r estimate}
estimate_result <- simmethods::POWSC_estimation(ref_data = ref_data,
                                                verbose = T,
                                                seed = 10)
```

## Simulating datasets using POWSC

After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.

1. Datasets with default parameters
2. Determin the number of cells
3. Simulate two groups


### Datasets with default parameters

The reference data contains 200 cells and 2000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data. In addtion, the simulated dataset will have one group of cells.

```{r default_params}
simulate_result <- simmethods::POWSC_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "SCE",
  seed = 111
)
```

```{r}
SCE_result <- simulate_result[["simulate_result"]]
dim(SCE_result)
```

```{r}
head(colData(SCE_result))
```

```{r}
head(rowData(SCE_result))
```


### Determin the number of cells

In POWSC, we can set `nCells` directly. For example, if we want to simulate 500 cells, we can type `other_prior = list(nCells = 500)`.
Here, we simulate a new dataset with 500 cells:
```{r customed cell and gene number}
simulate_result <- simmethods::POWSC_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(nCells = 500),
  seed = 111
)
```

```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```


### Simulate two groups

POWSC will automatically simulate two cell groups by default. Users can set `de.prob` to specify the proportion of DEGs between two groups.

```{r groups}
simulate_result <- simmethods::POWSC_simulation(
  parameters = estimate_result[["estimate_result"]],
  return_format = "list",
  other_prior = list(nCells = 500,
                     de.prob = 0.2),
  seed = 111
)
```

```{r}
result <- simulate_result[["simulate_result"]][["count_data"]]
dim(result)
```


```{r}
## cell information
cell_info <- simulate_result[["simulate_result"]][["col_meta"]]
table(cell_info$group)
```

```{r}
## gene information
gene_info <- simulate_result[["simulate_result"]][["row_meta"]]
### the proportion of DEGs
table(gene_info$de_gene)[2]/nrow(result) ## de.prob = 0.2
```
