---
title: "Kersplat"
output:
  blogdown::html_page:
    toc: yes
weight: 3
---


<div id="TOC">
<ul>
<li><a href="#estimating-parameters-from-a-real-dataset" id="toc-estimating-parameters-from-a-real-dataset">Estimating parameters from a real dataset</a></li>
<li><a href="#simulating-datasets-using-kersplat" id="toc-simulating-datasets-using-kersplat">Simulating datasets using Kersplat</a>
<ul>
<li><a href="#datasets-with-default-parameters" id="toc-datasets-with-default-parameters">Datasets with default parameters</a></li>
<li><a href="#determin-the-number-of-cells-and-genes" id="toc-determin-the-number-of-cells-and-genes">Determin the number of cells and genes</a></li>
</ul></li>
</ul>
</div>

<!-- github markdown built using 
rmarkdown::render("vignettes/Kersplat.Rmd", output_format = rmarkdown::github_document())
-->
<p>Here Kersplat method will be demonstrated clearly and hope that this document can help you.</p>
<div id="estimating-parameters-from-a-real-dataset" class="section level2">
<h2>Estimating parameters from a real dataset</h2>
<p>Before simulating datasets, it is important to estimate some essential parameters from a real dataset in order to make the simulated data more real. If you do not have a single-cell transcriptomics count matrix now, you can use the data collected in simmethods package by <code>simmethods:data</code> command.</p>
<pre class="r"><code>library(simmethods)
library(SingleCellExperiment)
# Load data
ref_data &lt;- simmethods::data
dim(ref_data)
# [1] 4000  160</code></pre>
<p>Using <code>simmethods::Kersplat_estimation</code> command to execute the estimation step.</p>
<pre class="r"><code>estimate_result &lt;- simmethods::Kersplat_estimation(ref_data = ref_data,
                                                   verbose = T,
                                                   seed = 10)
# Estimating parameters using Kersplat
# Warning in newKersplatParams(): The Kersplat simulation is still experimental
# and may produce unreliable results. Please try it and report any issues to
# https://github.com/Oshlack/splatter/issues. The development version may have
# improved features.
# Raw: 0.180467969462491 A: 5.24644589782513 B: 1.56749149061734 C: -3.99991188833989 Y: 0.796802242740237
# Warning in kersplatEstBCV(counts, params, verbose): Exponential corrected BCV is
# negative.Using linear correction.
# Warning in kersplatEstBCV(counts, params, verbose): Linear corrected BCV is
# negative.Using existing bcv.common.</code></pre>
</div>
<div id="simulating-datasets-using-kersplat" class="section level2">
<h2>Simulating datasets using Kersplat</h2>
<p>After estimating parameter from a real dataset, we will simulate a dataset based on the learned parameters with different scenarios.</p>
<ol style="list-style-type: decimal">
<li>Datasets with default parameters</li>
<li>Determin the number of cells and genes</li>
</ol>
<div id="datasets-with-default-parameters" class="section level3">
<h3>Datasets with default parameters</h3>
<p>The reference data contains 160 cells and 4000 genes, if we simulate datasets with default parameters and then we will obtain a new data which has the same size as the reference data.</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Kersplat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;SCE&quot;,
  seed = 111
)
# nCells: 160
# nGenes: 4000</code></pre>
<pre class="r"><code>SCE_result &lt;- simulate_result[[&quot;simulate_result&quot;]]
dim(SCE_result)
# [1] 4000  160</code></pre>
<pre class="r"><code>head(colData(SCE_result))
# DataFrame with 6 rows and 1 column
#         cell_name
#       &lt;character&gt;
# Cell1       Cell1
# Cell2       Cell2
# Cell3       Cell3
# Cell4       Cell4
# Cell5       Cell5
# Cell6       Cell6</code></pre>
</div>
<div id="determin-the-number-of-cells-and-genes" class="section level3">
<h3>Determin the number of cells and genes</h3>
<p>Here, we simulate a new dataset with 500 cells and 1000 genes:</p>
<pre class="r"><code>simulate_result &lt;- simmethods::Kersplat_simulation(
  parameters = estimate_result[[&quot;estimate_result&quot;]],
  return_format = &quot;list&quot;,
  other_prior = list(nCells = 500,
                     nGenes = 1000),
  seed = 111
)
# nCells: 500
# nGenes: 1000</code></pre>
<pre class="r"><code>result &lt;- simulate_result[[&quot;simulate_result&quot;]][[&quot;count_data&quot;]]
dim(result)
# [1] 1000  500</code></pre>
</div>
</div>
