---
title: "Data Simulation"
output:
  blogdown::html_page:
    toc: true
weight: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#", collapse = TRUE)
```

We have already known how to estimate parameters from one or more real datasets and got the estimation results. In this chapter, we will demonstrate how to simulate single-cell transcriptomics data based on the previous estimation results, especially the useful parameters that are usually customized to satisfy the different application situations.

For demonstrations, we use **Splat** method as it contains all functionalities and available parameters that we want to introduce.

Library our packages first:

```{r library_packages, message=FALSE}
library(simmethods)
library(simpipe)
```


## Simulate Datasets From Local R

### Step1: Prepare Estimation Results

Load data and perform estimation:

```{r data_load_estimation, warning=FALSE}
ref_data <- simmethods::data
estimation_result <- simmethods::Splat_estimation(
  ref_data = ref_data,
  verbose = TRUE,
  seed = 666
)
```

### Step2: Check Availabel Parameters

Next, check the optional parameters that control the size of the simulated datasets, the proportion of DEGs, the number of cell batches and datasets with cellular trajectory. In this way, you will know the essential parameters that may satisfy your simulation requirements.

```{r parameters_check, eval=FALSE}
help(SplatPop_simulation)
```

```{r}
## Details
# In addtion to simulate datasets with default parameters, users want to simulate other kinds of datasets, e.g. a counts matrix with 2 or more cell groups. In Splat, you can set extra parameters to simulate datasets.
# 
# The customed parameters you can set are below:
# 
# nCells. In Splat, you can not set nCells directly and should set batchCells instead. For example, if you want to simulate 1000 cells, you can type other_prior = list(batchCells = 1000). If you type other_prior = list(batchCells = c(500, 500)), the simulated data will have two batches.
# 
# nGenes. You can directly set other_prior = list(nGenes = 5000) to simulate 5000 genes.
# 
# nGroups. You can not directly set other_prior = list(nGroups = 3) to simulate 3 groups. Instead, you should set other_prior = list(prob.group = c(0.2, 0.3, 0.5)) where the sum of group probabilities must equal to 1.
# 
# de.prob. You can directly set other_prior = list(de.prob = 0.2) to simulate DEGs that account for 20 percent of all genes.
# 
# prob.group. You can directly set other_prior = list(prob.group = c(0.2, 0.3, 0.5)) to assign three proportions of cell groups. Note that the number of groups always equals to the length of the vector.
# 
# nBatches. You can not directly set other_prior = list(nBatches = 3) to simulate 3 batches. Instead, you should set other_prior = list(batchCells = c(500, 500, 500)) to reach the goal and the total cells are 1500.
# 
# If users want to simulate datasets for trajectory inference, just set other_prior = list(paths = TRUE). Simulating trajectory datasets can also specify the parameters of group and batch. See Examples.
```

These parameters can be categorized into 4 classes and respectively represent the main four functionalities in Splat method:

- parameters for cell groups

- parameters for DEGs

- parameters for batches

- parameters for cellular differentiation trajectory

In the next part of step3, we will describe these application situations in detail.


### Step3: Simulation

#### Task1: The Number of Cell and Gene

The first application situation is generating datasets with different number of cells and genes. After browsing the vignettes of Splat method, we know that `batchCells` parameter controls the number of cells and `nGenes` controls the number of genes.

Simulate `1000` cells and `5000` genes:

```{r 1000_5000}
data_1000_5000 <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 5000),
  return_format = "Seurat",
  verbose = TRUE,
  seed = 666
)
```

```{r check_1000_5000}
data_1000_5000$simulate_result
```

Simulate `10000` cells and `20000` genes:

```{r 10000_20000}
data_10000_20000 <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 10000,
                     nGenes = 20000),
  return_format = "Seurat",
  verbose = TRUE,
  seed = 666
)
```

See the number of cells and genes
```{r check_10000_20000}
data_10000_20000$simulate_result
```

Check the execution time:
```{r time_10000_20000}
data_10000_20000$simulate_detection$Elapsed_Time_sec
```


#### Task2: Cell Groups

If we want to simulate two groups of cells using Splat method, we can use `prob.group` parameter to specify the proportions of cells in two groups. The length of `prob.group` vector defines the number of groups.

Simulate two groups (4:6):
```{r 4:6}
data_4_6 <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 5000,
                     prob.group = c(0.4, 0.6)),
  return_format = "Seurat",
  verbose = TRUE,
  seed = 666
)
```

Check group labels of cells
```{r check_4_6}
table(data_4_6$simulate_result$group)
```


Simulate five groups (1:1:2:3:3):
```{r 11233}
data_11233 <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 5000,
                     prob.group = c(0.1, 0.1, 0.2, 0.3, 0.3)),
  return_format = "Seurat",
  verbose = TRUE,
  seed = 666
)
```

Check group labels of cells
```{r check_11233}
table(data_11233$simulate_result$group)
```


#### Task3: Differential Expressed Genes

Users can also set the proportion of DEGs in Splat method via `de.prob` parameter which ranges from 0 to 1.

Here we set `de.prob` as 0.2 to simulate 20% DEGs in two cell groups.

```{r 2_0.2}
simulated_data <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 5000,
                     prob.group = c(0.4, 0.6),
                     de.prob = 0.2),
  return_format = "list",
  verbose = TRUE,
  seed = 666
)
```

Check group labels of cells
```{r check_2_0.2}
table(simulated_data$simulate_result$col_meta$group)
```

Check the proportion of DEGs
```{r check_2_0.2_pro}
row_meta <- simulated_data$simulate_result$row_meta
table(row_meta$de_gene == "yes")/length(row_meta$de_gene)
```

We then simulate another dataset which contains more than 2 groups (4 groups and 40% DEGs):

```{r 4_0.4}
simulated_data <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 5000,
                     prob.group = c(0.2, 0.2, 0.3, 0.3),
                     de.prob = 0.4),
  return_format = "list",
  verbose = TRUE,
  seed = 666
)
```

Check group labels of cells
```{r check_4_0.4}
table(simulated_data$simulate_result$col_meta$group)
```

Check the proportion of DEGs
```{r check_4_0.4_pro}
row_meta <- simulated_data$simulate_result$row_meta
table(row_meta$de_gene == "yes")/length(row_meta$de_gene)
```

Note that we can know the DEGs between any pair of two groups in Splat method (except for **scDesign** and **SPARSim**). For example, if we want to
get the DEGs between the group1 and group2, we should extract the `DEFactor` in gene metadata:

```{r DEFactor}
gene_meta <- simulated_data$simulate_result$row_meta
DEFactor1 <- gene_meta$DEFacGroup1
DEFactor2 <- gene_meta$DEFacGroup2
```

Then we do the division:

```{r DEGs}
DEFactor <- DEFactor1/DEFactor2
```

Check the gene that whose DEFactor is not equal to 1 and they are defined as the DEGs between group1 and group2:

```{r check_gene}
table(DEFactor != 1)
DEGs_group1_group2 <- rownames(gene_meta)[DEFactor != 1]
DEGs_group1_group2[1:10]
```

{{% notice warning %}}
scDesign and SPARSim can not return the DEGs between any pair of groups when the number of cell groups is higher than 2. But when there are only two groups in a simulated data, the DEGs are valid.
{{% /notice %}}


#### Task4: Cell Batches

Simulating different cell batches is also an important application situation in many researches related to benchmarking and method development.

In Splat and many other methods, users can specify the number of cell batches and the cell numbers in every batch via `batchCells` parameter. Here, we will simulate 3 batches with cell numbers of 1000, 2000 and 3000, respectively.

```{r batch}
simulated_data <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = c(1000, 2000, 3000),
                     nGenes = 5000),
  return_format = "list",
  verbose = TRUE,
  seed = 666
)
```

Check the batches:
```{r check_batches}
table(simulated_data$simulate_result$col_meta$batch)
```


#### Task5: Cellular Trajectory

Using Splat method to simulate the data with cellular differentiation trajectory is another application situation of data simulation. Simply, we can set `paths` parameter as `TRUE`.

```{r paths_TRUE}
simulated_data <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     prob.group = c(0.3, 0.2, 0.5),
                     nGenes = 5000,
                     paths = TRUE),
  return_format = "SingleCellExperiment",
  verbose = TRUE,
  seed = 666
)
```

```{r plot_path}
library(scater)
sim.paths <- logNormCounts(simulated_data$simulate_result)
sim.paths <- runPCA(sim.paths)
plotPCA(sim.paths, colour_by = "group")
```

If you want to set other parameters related to the trajectory in Splat method, you can browse the official vignettes represented in `Splatter` package and the [website](https://www.bioconductor.org/packages/release/bioc/vignettes/splatter/inst/doc/splat_params.html).

```{r splat_vignetts, eval=FALSE}
help(splatSimulate, package = "splatter")
```

Here, we only add extra two parameters `path.nSteps` and `path.skew`:

```{r custome_path}
simulated_data <- simmethods::Splat_simulation(
  parameters = estimation_result$estimate_result,
  other_prior = list(batchCells = 1000,
                     prob.group = c(0.3, 0.2, 0.5),
                     nGenes = 5000,
                     paths = TRUE,
                     path.nSteps = 20,
                     path.skew = 0.1),
  return_format = "SingleCellExperiment",
  verbose = TRUE,
  seed = 666
)
```

```{r plot_path2}
library(scater)
sim.paths <- logNormCounts(simulated_data$simulate_result)
sim.paths <- runPCA(sim.paths)
plotPCA(sim.paths, colour_by = "group")
```

## Simulate Datasets From Docker in R

This part we will demonstrate how to simulate datasets by using Docker in R and users should make sure that [Docker](https://www.docker.com/) has been installed on your device.

First, start Docker and check:
```{r check_docker}
library(simpipe2docker)
test_docker_installation(detailed = TRUE)
```

Estimation parameters from Docker:

```{r estimation_docker, warning=FALSE}
estimation_result <- simpipe2docker::estimate_parameters_container(
  ref_data = ref_data,
  method = "Splat",
  verbose = TRUE,
  seed = 666
)
```

Simulate new datasets from Docker:

```{r docker_sim_1, warning=FALSE}
## simulate 1000 cells and 1000 genes
simulated_data <- simpipe2docker::simulate_datasets_container(
  parameters = estimation_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 1000),
  return_format = "SingleCellExperiment",
  verbose = TRUE,
  seed = 666
)
simulated_data$refdata_Splat_1$simulate_result
```


```{r docker_sim_2, warning=FALSE}
## simulate 1000 cells and 1000 genes (two groups and 40% DEGs)
simulated_data <- simpipe2docker::simulate_datasets_container(
  parameters = estimation_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 1000,
                     prob.group = c(0.4, 0.6),
                     de.prob = 0.4),
  return_format = "list",
  verbose = TRUE,
  seed = 666
)
```

## Simulate Datasets By Simpipe Package

Based on **simmethods** package, **simpipe** package provides the other useful functions. Users can estimate parameters from multiple real datasets by using multiple methods. Meanwhile, users can also simulate multiple new datasets at once. In this part, we introduce some helpful functions in **simpipe** package.

First, we should use simpipe to estimate parameters from two real datasets:

```{r simpipe_est, warning=FALSE}
## prepare a list of data
data <- list(data1 = ref_data,
             data2 = ref_data)

estimation_result <- simpipe::estimate_parameters(
  ref_data = data,
  method = "Splat",
  verbose = TRUE,
  seed = 666
)
```


### Generate Multiple Datasets For Every Estimation Result

For every estimation result, we can generate multiple datasets by setting `n` parameter in `simulate_datasets` function:

```{r simpipe_n}
simulated_data <- simpipe::simulate_datasets(
  parameters = estimation_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 1000),
  n = 3,
  return_format = "list",
  verbose = TRUE,
  seed = 666
)
```

We can also set `seed` parameter whose length is equal to the number of `n`:

```{r simpipe_n2}
simulated_data <- simpipe::simulate_datasets(
  parameters = estimation_result,
  other_prior = list(batchCells = 1000,
                     nGenes = 1000),
  n = 3,
  return_format = "list",
  verbose = TRUE,
  seed = c(666, 888, 999)
)
```
